src/sync.h

```diff
@@ -400,4 +398,104 @@ class CSemaphoreGrant
     }
 };
 
+/**
+ * A smart container that protects an owned object with a mutex.
+ * This is like a mutex with an extension - owning an object of arbitrary type
+ * and providing a dereference operation which gives access to the owned object
+ * and transparently locks the mutex for the duration of the access.
+ * For example, the following is safe:
+ * @code{.cpp}
+ * Synced<std::vector<int>> v;
+ *
+ * std::thread t1{[&v]() { v->push_back(5); }};
+ * std::thread t2{[&v]() { v->push_back(6); }};
+ * @endcode
+ *
+ * It is possible so acquire the mutex for longer time than just for a single call:
+ * @code{.cpp}
+ * Synced<std::vector<int>> v{5, 6, 7, 8};
+ *
+ * // v is locked for the entire duration of the loop
+ * for (auto& i : *v) {
+ *     i *= 10;
+ * }
+ *
+ * {
+ *     LOCK(v);
+ *     size_t size = v->size();
+ *     // The container will not be changed by another thread between the two size() calls.
+ *     assert(size == v->size());
+ * }
+ * @endcode
+ */
+template <typename T>
+class Synced : public RecursiveMutex
```

Comment:
I don't think we should be doing more recursive mutex stuff (they're harder to reason about, and less efficient any time that matters). I haven't been able to figure out a way of doing this sensibly otherwise, however -- clang just gets too confused about aliasing.

Reply:
It used `Mutex` before and mimicked a `RecursiveMutex`, but I totally agree with @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/25390#discussion_r919129254) and changed it to `RecursiveMutex`.

Even before that there was another approach which did not need a recursive mutex, and double locking was not allowed, but [you did not like that](https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1158434634).

Reply:
In my opinion: we shouldn't allow double locking (for almost all locks), and should give compile time errors when double locking is attempted. I'd love to see a variant of this that achieved both those things.

Reply:
Done - in the latest incarnation no double locking is allowed.